### ISSUE TITLE
BRouter integration: Migrate access of routing tile data to SAF

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#10176]
Routing tile data (files with extension '.rd5') reside in the `segments4` subfolder. As they tend to be pretty large (up to 150 MB and more), and as there are many of them potentially, they should be sharable with other apps, e. g. Locus. Therefore accessing files via SAF is needed, and the current caching mechanism needs to be adapted. (See `PhysicalFile` etc.)

@eddiemuc 
This is the part we've talked about.

**Comment by eddiemuc** [comment]
@moving-bits after BRouter stuff is now in master, can/should I start working on this issue?

**Comment by moving-bits** [comment]
> @moving-bits after BRouter stuff is now in master, can/should I start working on this issue?

That would be great. Starting point is `brouter/BRouterService`.

**Comment by eddiemuc** [comment]
Hi @moving-bits 
I thought before I dive into filtering too deep I should do the SAF migration of BRouter before that :-)

I took a look at the code, before I walt too much in the wrong direction I want to verify that my understanding of what needs to be done is correct:

There shall be two new public folders:
  * PersistableFolder.ROUTING_BASE (location defaults to BASE/routing). This folder shall contain:
       * multiple "rawTrackPath" files (named with suffix _rawtrack.dat"
       * multiple "profile" files (named e.g. fastbike.brf"
       * a file "lookups.dat" which you seem to have migrated to SAF already?
   * PersistableFolder.ROUTING_TILES (defaults to ROUTING_BASE/segment4). This folder contains multiple files with tile data inside (with suffix .rd5)

Is that about it or am I missing something?

I also noticed that class RouterEngine has a suspicious members "outfileBase" and "logfileBase", which however seems to be initialized to null always. Is this intended, can the be removed?

**Comment by Lineflyer** [comment]
I think there is no need for two separate folders. Either all data could be in/below a single folder, or the non user modifiable data could be stored internally.

**Comment by eddiemuc** [comment]
> I think there is no need for two separate folders. Either all data could be in/below a single folder, or the non user modifiable data could be stored internally.
> 
> 

Iirc the idea was to separate the data sharable between applications (the tile data, which is large) from the cgeo-specific data (the other data e.g. profiles). Just because there are two public folders does.not mean that both are settable by the user. We have many public folders not settable by the user e.g. logfiles or spoiler-image-dir

**Comment by Lineflyer** [comment]
I though we have one place/folder structire with tile data and routing profiles (as AFAIK this is the common base of external and internal) and the rest could just be internal without any SAF or user selectable folders.

**Comment by moving-bits** [comment]
Hi @eddiemuc,

thanks for diving into this.

> * PersistableFolder.ROUTING_BASE (location defaults to BASE/routing). This folder shall contain:
>       
>   * multiple "rawTrackPath" files (named with suffix _rawtrack.dat"

I'm not sure about this one. It looks to me like some internal caching files, so they may be placed into app-internal storage better, probably, and continued to being accessed using `File`.

>  * multiple "profile" files (named e.g. fastbike.brf"

Correct. Those files will be checked for at starting internal BRouter, and created from app resources, if necessary, automatically.
Those files may be updated/changed/replaced by users, so I placed them into an accessible folder. But they are small enough that I did not see the need here for a configurable folder for sharing those files with other apps.

>  * a file "lookups.dat" which you seem to have migrated to SAF already?

Yes, same as above. In this case "migrated" already. But maybe we can get rid of this file completely by either providing it internally from app resources only, or by placing this file into an app-internal folder.

> * PersistableFolder.ROUTING_TILES (defaults to ROUTING_BASE/segment4). This folder contains multiple files with tile data inside (with suffix .rd5)

Exactly. This should stay a separate, configurable public folder to be able to share those rather large files with other apps.

> Is that about it or am I missing something?

No, that should be the complete list. I've removed all other files and folder from the original BRouter folder structure already.

> I also noticed that class RouterEngine has a suspicious members "outfileBase" and "logfileBase", which however seems to be initialized to null always. Is this intended, can the be removed?

This can be removed probably. There is (has been) lots of logging and/or tracing code, much of which I removed already, at least the more obvious parts. But it's very likely that even more unused code can be removed. Also on non-logging methods, but I've been a bit reluctant here as I'm unsure which parts may still be needed as soon as we start using voice hints / turn-by-turn instructions. I even had to reinsert some code already to be able to run the tests.

**Comment by moving-bits** [comment]
> I think there is no need for two separate folders. Either all data could be in/below a single folder, or the non user modifiable data could be stored internally.

Theoretically yes, but in case of BRouter the first folder contains some small, but user-configurable files, and the other contains the non-user-modifiable but large files, which are candidates to be shared. But the user will only have to select the second folder (and the default points to a reasonable c:geo-specific value). The first folder is SAF, but in a fixed place relative to c:geo BASE folder - so no need for user-interaction here.

**Comment by moving-bits** [comment]
> I though we have one place/folder structire with tile data and routing profiles (as AFAIK this is the common base of external and internal) and the rest could just be internal without any SAF or user selectable folders.

I've adapted the original BRouter folder structure to minimize it, and also removed all folders and files not needed for running the integrated routing engine. Three types of files remained:
1) routing tile data => large files, non-user-modifiable, probably to be shared with other apps (even possible with external BRouter)
2) routing profile files => a couple of small files, user-modifiable, but in a separate folder in both external and internal BRouter
3) `lookups.dat` => small config file, non-user-modifiable - need to investigate a bit further, if we can eliminate having this file in a public folder at all

Only 1) needs a user-selectable folder to be able to share it; must be SAF
For 2) we need a public SAF-aware folder to allow the user editing this file; folder location is fixed
Currently 3) is in the same folder as 2, no additional user-interaction for both

**Comment by eddiemuc** [comment]
Thanks a lot, I think I have everything I need to get started

**Comment by eddiemuc** [comment]
@moving-bits while working on this I had problems Debugging into the BRouter code. Somehow my breakpoints were not hit when set in it. Other breakpoints (in "normal" c:geo code) were hit without problems.

I found indication that also the internal BRouter is somehow started as a separate service, at some point Android Studio presented me a separate"BRouter" process to attach my debugger in addition to the main c:geo process. However this did not help either, breakpoints were still not hit.

I have the suspicion that somehow a BRouter service/process is recreated each time the map viewer is opened (and thus even if I attached debugger to a previous one this is lost because the process is created anew). Is this somehow correct?

And do you have any recommendations on how to debug into BRouter code?

**Comment by moving-bits** [comment]
Yes, the integrated BRouter is still started as a service - I wanted to keep things (especially the interface) as much the same as possible, at least in the first step. (We may decide to change a couple of things later, when the basic integration is completed. This may regard calling BRouter as a service, transferring data via GPX instead of Geopoints etc.)

As for debugging: This has worked for me, but, as you already mentioned, you have to attach the debugger to the BRouter process. Can't remember if that connection got lost on closing/reopening the map, though - maybe that's a use case that did not occur during my debugging sessions on BRouter. (And can't really test currently, either, as I'm not at my dev machine, sorry.)

**Comment by eddiemuc** [comment]
HI @moving-bits I have problems getting the existing (legacy, non-saf-migrated) code running. Using downloaded tiles it keeps telling me that calculated checksums (CRC) of the tiles is wrong. Any idea what could be the reason? How can I get valid routing tile data which with the legacy code is running smoothly?

**Comment by moving-bits** [comment]
Hm, I'm unsure with that. Haven't looked up how the checksums are calculated. 

One idea: There are several method with non-final parameters. During migration I made sure to not change those parameters which get modified within the method - but I haven't checked multi-level changes (and the compiler did not complain).

Have you binary-compared tile data downloaded with the downloader to same file downloaded with BRouter? Or to the same file downloaded manually from the website? (Only to make sure whether it's corrupted data or the checksum calculation being defect.)

**Comment by eddiemuc** [comment]
@moving-bits #10323 introduces the FileByteReader class which should be the missing link to migrate BRouter to SAF (it "replaces" RandomFileAccess with something working on FileInoutStream)

Feel free to dive in from here, I can't really continue until the basics work.

**Comment by moving-bits** [comment]
Hi eddiemuc,

thanks for diving into this. I've just tried to reproduce the checksum error, but don't see it.

> HI @moving-bits I have problems getting the existing (legacy, non-saf-migrated) code running. Using downloaded tiles it keeps telling me that calculated checksums (CRC) of the tiles is wrong. Any idea what could be the reason? How can I get valid routing tile data which with the legacy code is running smoothly?

I did the following to test:
- Settings: internal routing, auto tile download active
- move the map to a place where I have no tile data for currently
- create a new individual route
- c:geo asks me whether to download the missing tile data, which I confirm
- after successful download do something to make the route recalculate (like adding another waypoint, change routing mode, close+open map etc.)

Everything's fine here, then. A reasonable looking route gets displayed.

Where do you see/get the CRC errors?

**Comment by eddiemuc** [comment]
@moving-bits I have to apologize, I forgot to continue working on this...
I will try to look into this next time I find a free time slot for c:geo work

**Comment by eddiemuc** [comment]
@moving-bits following up on this: I downloaded the tile data directly from http://brouter.de/brouter/segments4/ as you suggested, and indeed the downloaded data differed significantly from the data sownloaded via c:geo-internal downloader AND it's checksum was correct!

Comparing the two files (downloaded by c:geo with downloaded by me today) shows changed, but since tiles are updated daily this does not say a lot. Only pointer I have is that the c:geo-internally downloaded file is a bit shorter than the other (1MB shorter, total file size is about 123MB) - and I don't erxpect that the map has changed significant recently. So my guess is that download was somehow truncated near the end in c:geo-internal download.

Three questions:
* Would it make sense to perform a validity check on downloaded tile data before making it available to the end user (e.g. by calculating said checksum)?
* Should I now continue working on this or did you already take over? (Sorry again for my forgetfullness...)
* For developer testing (since I am not using routing in c:geo myself): what is a good way to reliably retrigger route calculation from within map viewer? I currently created an individual route and do a resort of this route for this, but is there an easier way?

For reference: I currently use Hamburg inner city for my tests, relevant routing tiles are E5_N50.rd5 and E10_N50.rd5

**Comment by fm-sys** [comment]
> So my guess is that download was somehow truncated near the end in c:geo-internal download.

That sounds a bit like "truncation on history trail export" (#10331) (#10376). Could it be related?

**Comment by eddiemuc** [comment]
PR #10404 will migrate brouter tile access (br5 files previously located in segment4) to SAF.

It does not add checksum check to download as asked above.

**Comment by eddiemuc** [comment]
Update: PR #10404 will now also provide the statis method `PhysicalFile.checkTileDataIntegrity` which allows for checking a tile file's integrity.

**Comment by moving-bits** [comment]
> Comparing the two files (downloaded by c:geo with downloaded by me today) shows changed [...] So my guess is that download was somehow truncated near the end in c:geo-internal download.

This is truly strange, as I use the system's built-in download service for downloading the tile data. Only after having received the file I copy it manually to the target destination. So a bug may lay there (if any).

> Would it make sense to perform a validity check on downloaded tile data before making it available to the end user (e.g. by calculating said checksum)?

Doesn't hurt at least (except for some delay). We need to look how/where to integrate into the download / copy / followup-action cycle, especially with the upcoming "copy under the hood in the background" functionality. So I would propose to open a separate issue for it, and wait with implementation until the copying service is implemented.

> Should I now continue working on this or did you already take over? (Sorry again for my forgetfullness...)

I haven't been online today until now, but as it seems, you already have continued and finished the task ;-) Thanks for it!

> For developer testing (since I am not using routing in c:geo myself): what is a good way to reliably retrigger route calculation from within map viewer? I currently created an individual route and do a resort of this route for this, but is there an easier way?

Having an individual route is a good starting point, as it gets persisted. Now to making it recalculate the easiest way is to change the route type using the map quick settings icon, e. g. change it from "walk" to "car" or vv. (Another option would be to add or remove a waypoint from the individual route, but that does not recalculate the whole route but only the affected segments.)

**Comment by moving-bits** [comment]
> > So my guess is that download was somehow truncated near the end in c:geo-internal download.
> 
> That sounds a bit like "truncation on history trail export" (#10331) (#10376). Could it be related?

Not sure. At least the output stream uses explicit resource management, see `downloader/ReceiveDownloadActivity:doCopy()`

**Comment by moving-bits** [comment]
The original issue is implemented with #10404, and for the "compare checksum after download" I've created a new issue #10416, so I guess we can close this issue. Thanks @eddiemuc for your support on this one!
